trinity/protocol/common/managers.py

```diff
@@ -0,0 +1,154 @@
+from abc import abstractmethod
+import asyncio
+from typing import (
+    cast,
+    Generic,
+    TypeVar,
+    Set,
+    Tuple,
+    Type,
+)
+
+from cancel_token import CancelToken
+
+from p2p.exceptions import (
+    ValidationError,
+)
+from p2p.peer import BasePeer, PeerSubscriber
+from p2p.protocol import (
+    Command,
+)
+from p2p.service import BaseService
+
+from trinity.exceptions import AlreadyWaiting
+
+from .requests import BaseRequest
+
+
+PeerClass = TypeVar('PeerClass', bound=BasePeer)
+RequestType = TypeVar('RequestType', bound=BaseRequest)
+ResponseType = TypeVar('ResponseType')
+ReturnType = TypeVar('ReturnType')
+
+
+class BaseRequestManager(PeerSubscriber, BaseService, Generic[PeerClass, RequestType, ResponseType, ReturnType]):  # noqa: E501
+    #
+    # PeerSubscriber
+    #
+    @property
+    def subscription_msg_types(self) -> Set[Type[Command]]:
+        return {self._response_msg_type}
+
+    msg_queue_maxsize = 100
+
+    response_timout: int = 60
+
+    pending_request: Tuple[RequestType, 'asyncio.Future[ReturnType]'] = None
+
+    _peer: PeerClass
+
+    def __init__(self, peer: PeerClass, token: CancelToken) -> None:
+        self._peer = peer
+        super().__init__(token)
+
+    #
+    # Service API
+    #
+    async def _run(self) -> None:
+        self.logger.debug("Launching %s for peer %s", self.__class__.__name__, self._peer)
+
+        with self.subscribe_peer(self._peer):
+            while self.is_running:
+                peer, cmd, msg = await self.wait(
+                    self.msg_queue.get(), token=self.cancel_token)
+                if peer != self._peer:
+                    self.logger.error("Unexpected peer: %s  expected: %s", peer, self._peer)
+                    continue
+                elif isinstance(cmd, self._response_msg_type):
+                    self._handle_msg(cast(ResponseType, msg))
+                else:
+                    self.logger.warning("Unexpected message type: %s", cmd.__class__.__name__)
+
+    async def _cleanup(self) -> None:
+        pass
+
+    def _handle_msg(self, msg: ResponseType) -> None:
+        if self.pending_request is None:
+            self.logger.debug(
+                "Got unexpected %s message from %", self.response_msg_name, self._peer
+            )
+            return
+
+        request, future = self.pending_request
+
+        try:
+            request.validate_response(msg)
+        except ValidationError as err:
+            self.logger.debug(
+                "Response validation failure for pending %s request from peer %s: %s",
+                self.response_msg_name,
+                self._peer,
+                err,
+            )
+        else:
+            future.set_result(self._normalize_response(msg))
+            self.pending_request = None
+
+    @abstractmethod
+    def _normalize_response(self, msg: ResponseType) -> ReturnType:
+        pass
+
+    @abstractmethod
+    def __call__(self) -> ReturnType:  # type: ignore
```

Comment:
You can drop that `ignore` here. Ignoring the violating places is enough. However, maybe we can get rid of the violations altogether. I'll elaborate in another comment.